### PR TITLE
fix pre-allocating slice capacity in pilot

### PR DESCRIPTION
**Pull request by ayj**

### CODE REVIEW

**Comment by ayj** [comment]
One observed effect of this bug is inaccurate logging. The following `info` message prints the number of virtual services `<B>` which is logged as 2x the number of actual virtual services.

```
XDS: Registry event, pushing. Services: <A>, VirtualServices: <B>, ConnectedEndpoints: <C>
```

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -158,7 +158,7 @@ func translateVirtualHost(
 		serviceByPort[80] = nil
 	}
 
-	out := make([]GuardedHost, len(serviceByPort))
+	out := make([]GuardedHost, 0, len(serviceByPort))
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/route/route.go]: can we just get rid of fixed len and make it dynamic?

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 2]
```diff
@@ -163,7 +163,7 @@ func (s *DiscoveryServer) updateModel() {
 		log.Errorf("XDS: failed to update virtual services %v", err)
 	} else {
 		s.virtualServiceConfigs = vservices
-		s.virtualServices = make([]*networking.VirtualService, len(vservices))
+		s.virtualServices = make([]*networking.VirtualService, 0, len(vservices))
```
Comment (baodongli) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: wonder if the change is necessary since the size of the array is fixed (as in len(vservices)).

**Comment by rshriram** [comment]
Merging due to stuck cla bot

**Pull request merged**